### ISSUE TITLE
Move to contextual components

### ISSUE THREAD

**Issue by willviles** [kolybasov/ember-medium-editor#6]
The inspiration for this idea comes from [Ember Leaflet](https://github.com/miguelcobain/ember-leaflet) and the way it uses contextual components with [Ember Composability Tools](https://github.com/miguelcobain/ember-composability-tools) to interact with the Leaflet API. 

Check out Miguel Andrade's [DOMless rendering with Ember](https://www.youtube.com/watch?v=nkiZ_cWoQMk) presentation from So Ember 2017, and read the Ember Composability Tools docs for a deeper dive into how it works.

At present, configuring Medium Editor isn't very 'Ember'. It's just one component with a huge options hash and doesn't leverage the power of the expressive HTMLbars templating or Ember actions.

Instead, I propose we should be able to build Medium Editor components more expressively using the type of templating suggested below.

## Note

@kolybasov, if you're interested in moving the addon in this direction, I'll shoulder a lot of the work, using Ember Leaflet's implementation as inspiration. I've got the next few days free to get this done, so if you've got some time to review & give feedback ASAP, then that'd be great!

### Default configuration 

Despite the move to contextual components, blockless definition would still work as expected.

```handlebars
{{medium-editor 
    model.text 
    options=(hash) 
    onChange=(action (mut model.text))}}
```

It would just be equivalent to this template.

```handlebars
{{#medium-editor 
    model.text 
    options=(hash) 
    onChange=(action (mut model.text)) 
    as | editor |}}
    
  {{#editor.toolbar as | buttons |}}
    {{buttons.bold}}
    {{buttons.italic}}
    {{buttons.underline}}
    {{buttons.anchor}}
    {{buttons.h1}}
    {{buttons.h2}}
    {{buttons.h3}}
    {{buttons.quote}}
  {{/editor.toolbar}}
  
  {{editor.anchor}}
  
{{/medium-editor}}
```

### Complex Configuration 

Below is an example of how full configuration of all toolbars, popups and plugins would look.

```handlebars
{{#medium-editor model.text options=(hash) onChange=(action (mut model.text)) as | editor |}}
  <!--
    Configure the editor toolbar's options and buttons.
    Ref: https://github.com/yabwe/medium-editor#button-options
  -->
  {{#editor.toolbar options=(hash) as | buttons |}}
    <!--
      Define buttons in desired order
    -->
    {{buttons.bold}}
    {{buttons.italic}}
    <!--
      Change the action, aria, tagNames, contentDefault, classList, attrs etc via options hash
    -->
    {{buttons.h1 options=(hash)}}
    <!--
      Alternatively, define the button's `contentDefault` via the yield
    -->
    {{#buttons.h2}}
      {{icon 'my-custom-h2-icon'}}
    {{/buttons.h2}}
    <!--
      Add custom editor toolbar buttons using the a custom-button component,
      wrapping the medium-button plugin. Define the `label`, `start` and `end` values,
      or handle via an onClick action.
      Ref: https://github.com/arcs-/medium-button#html-buttons
    -->
    {{buttons.custom options=(hash) onClick=(action)}}
    <!--
      Components can extend the base custom-button component and be used like so
    -->
    {{my-custom-button-component}}

  {{/editor.toolbar}}

  <!--
    Treat the anchor popup as a separate component. Allow it to be configured by an options hash.
    Ref: https://github.com/yabwe/medium-editor#anchor-form-options
  -->
  {{editor.anchor options=(hash)}}

  <!--
    Configure the insert plugin via a contextual component.
    The insert plugin by default enables insertion of images and embeds from social media.
    Both content types can be styled from a style toolbar and interacted with via an actions toolbar.
    Both toolbars can have buttons registered to them via contextual component declarations.
    Ref: https://github.com/orthes/medium-editor-insert-plugin
  -->
  {{#editor.insert as | insert |}}
    <!--
      Configure the insert plugin images via a contextual component
      Ref: https://github.com/orthes/medium-editor-insert-plugin/wiki/v2.x-Configuration#images-options
    -->
    {{#insert.images
        options=(hash)
        onUploadComplete=(action)
        onUploadFail=(action)
        as | image |}}
      <!--
        Define all the available style buttons for the image
      -->
      {{#image.styles as | buttons |}}
        <!--
          Define buttons in desired order
        -->
        {{buttons.left}}
        <!--
          Buttons expose their `add` and `remove` callbacks
        -->
        {{buttons.right onAdd=(action) onRemove=(action)}}
        <!--
          Define the button's `label` value via the yield
        -->
        {{#buttons.wide}}
          {{icon 'my-custom-wide-icon'}}
        {{/buttons.wide}}
        <!--
          Custom image styles can be created as components using a base component
        -->
        {{my-custom-image-style-button}}
      {{/image.styles}}
      <!--
        Define buttons inside the image interaction toolbar
      -->
      {{#image.toolbar as | buttons |}}
        <!--
          Define all the available buttons for the image toolbar
        -->
        {{#buttons.remove}}
          {{icon 'my-custom-remove-icon'}}
        {{/buttons.remove}}
        <!--
          Add custom image toolbar buttons.
        -->
        {{buttons.custom options=(hash) onClick=(action)}}
      {{/image.toolbar}}

    {{/insert.images}}

    <!--
      Configure the insert plugin embeds via a contextual component
      Ref: https://github.com/orthes/medium-editor-insert-plugin/wiki/v2.x-Configuration#embeds-options
    -->
    {{#insert.embeds
      options=(hash)
      as | embed |}}
      <!--
        Define all the available style buttons for the embed
      -->
      {{#embed.styles as | buttons |}}
        <!-- Works the same as {{#image.styles}} -->
      {{/embed.styles}}
      <!--
        Define buttons inside the embed interaction toolbar
      -->
      {{#embed.toolbar as | buttons |}}
        <!-- Works the same as {{#image.toolbar}} -->
      {{/embed.toolbar}}

    {{/insert.embeds}}

  {{/editor.insert}}

{{/medium-editor}}
```

**Comment by kolybasov** [comment]
@willviles sorry, don't have too much time for review. Actually looks amazing so if you think it is worth to be implemented then go ahead!

**Comment by willviles** [comment]
@kolybasov I'll go ahead and create an experimental branch!

It'll definitely involve merging Ember Medium Editor Button functionality into this addon and will probably include the Ember Medium Insert Plugin functionality too, as per https://github.com/kolybasov/ember-medium-editor/issues/3#issuecomment-307805731.

**Comment by n1ru4l** [comment]
Looks rad 👍

**Comment by joshuairl** [comment]
Excited for this!

**Comment by kolybasov** [comment]
Hey @willviles @n1ru4l @joshuairl! I've implemented contextual-components inside of #8.

And decided to drop some old API's:

- `onChange` is `onInput` now because I do not track if content is actually changed. It is basically raw `editableInput` but with 3 arguments: [text, InputEvent, medium-editor-dom-element]
- `onFinishedTyping` is deleted. And there are no any custom events now because the goal of this project is to be 1:1 wrapper(onInput exists to be able use typical ember pattern `onInput=(action (mut text))`)
- `options` hash is deleted. If you want to set some option do it directly on related component. For example `{{editor.toolbar static=true}}`

It may be not final API and a reason why I writing this message is feedback request. I would like to discuss any changes to current API or confirmation that this one is good. I am starting to write tests and docs. So if there will be no changes in about one week I will release first beta!
